### PR TITLE
ocamlPackages.js_of_ocaml*: 3.7.0 -> 3.8.0

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

Tried to see if I could do the broken auto-update properly. Unfortunately it seems that this update currently breaks
`eliom` which should be fixed as soon as they support ocaml-migrate-parsetree 2.1.0 or we have a patch for it:

```
+ ocamlfind ocamlc -c -g -keep-locs -w +A-4-6-7-9-40-42-44-48 -package 'ppx_tools_versioned, ppx_tools_versioned.metaquot_408' -I src/ppx -o src/ppx/ppx_eliom_client.cmi src/ppx/ppx_eliom_client.mli
findlib: [WARNING] Package ocaml-migrate-parsetree has multiple definitions in /nix/store/4fpj30ah88j7zcpc3mgrgwc4y4qmdkyl-ocaml4.10.0-ocaml-migrate-parsetree-2.1.0/lib/ocaml/4.10.0/site-lib/ocaml-migrate-parsetree/META, /nix/store/c0sga3bi7dbbh40lsfg9i91w9s9aj542-ocaml4.10.0-ocaml-migrate-parsetree-1.8.0/lib/ocaml/4.10.0/site-lib/ocaml-migrate-parsetree/META
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /nix/store/lqgcqn9nh1bknq3dj7qnh1k7fcapsdag-ocaml-4.10.0/lib/ocaml, /nix/store/lqgcqn9nh1bknq3dj7qnh1k7fcapsdag-ocaml-4.10.0/lib/ocaml/compiler-libs
File "src/ppx/ppx_eliom_client.mli", line 2, characters 45-78:
2 |   Migrate_parsetree.Versions.OCaml_408.types Migrate_parsetree.Driver.rewriter
                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Unbound module Migrate_parsetree.Driver
Command exited with code 2.
Compilation unsuccessful after building 11 targets (4 cached) in 00:00:00.
make: *** [Makefile:14: native] Error 10
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
@vbgl rebased against the janeStreet update seems like this fixed things, currently running a `nixpkgs-review` to make sure, but seems promising: `eliom` is no longer broken by this change.

**Comment by sternenseemann** [comment]
Result of `nixpkgs-review pr 106343` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>65 packages built:</summary>
  <ul>
    <li>acgtk</li>
    <li>dune-release</li>
    <li>eff</li>
    <li>framac</li>
    <li>jackline</li>
    <li>ocaml-crunch</li>
    <li>ocamlPackages.asn1-combinators</li>
    <li>ocamlPackages.async_js</li>
    <li>ocamlPackages.bonsai</li>
    <li>ocamlPackages.ca-certs</li>
    <li>ocamlPackages.cohttp-lwt-unix</li>
    <li>ocamlPackages.conduit-lwt-unix</li>
    <li>ocamlPackages.crunch</li>
    <li>ocamlPackages.curly</li>
    <li>ocamlPackages.dns</li>
    <li>ocamlPackages.dns-client</li>
    <li>ocamlPackages.eliom</li>
    <li>ocamlPackages.fiat-p256</li>
    <li>ocamlPackages.functoria</li>
    <li>ocamlPackages.functoria-runtime</li>
    <li>ocamlPackages.git</li>
    <li>ocamlPackages.git-http</li>
    <li>ocamlPackages.git-unix</li>
    <li>ocamlPackages.graphql-cohttp</li>
    <li>ocamlPackages.incr_dom</li>
    <li>ocamlPackages.index</li>
    <li>ocamlPackages.irmin-fs</li>
    <li>ocamlPackages.irmin-git</li>
    <li>ocamlPackages.irmin-graphql</li>
    <li>ocamlPackages.irmin-http</li>
    <li>ocamlPackages.irmin-mem</li>
    <li>ocamlPackages.irmin-pack</li>
    <li>ocamlPackages.irmin-test</li>
    <li>ocamlPackages.irmin-unix</li>
    <li>ocamlPackages.js_of_ocaml</li>
    <li>ocamlPackages.js_of_ocaml-compiler</li>
    <li>ocamlPackages.js_of_ocaml-lwt</li>
    <li>ocamlPackages.js_of_ocaml-ocamlbuild</li>
    <li>ocamlPackages.js_of_ocaml-ppx</li>
    <li>ocamlPackages.js_of_ocaml-ppx_deriving_json</li>
    <li>ocamlPackages.js_of_ocaml-tyxml</li>
    <li>ocamlPackages.metrics-unix</li>
    <li>ocamlPackages.mirage</li>
    <li>ocamlPackages.mirage-crypto-pk</li>
    <li>ocamlPackages.mirage-crypto-rng</li>
    <li>ocamlPackages.mirage-crypto-rng-mirage</li>
    <li>ocamlPackages.mirage-logs</li>
    <li>ocamlPackages.mirage-runtime</li>
    <li>ocamlPackages.mirage-unix</li>
    <li>ocamlPackages.mtime</li>
    <li>ocamlPackages.ocplib-json-typed-browser</li>
    <li>ocamlPackages.ocsigen-start</li>
    <li>ocamlPackages.ocsigen-toolkit</li>
    <li>ocamlPackages.opium</li>
    <li>ocamlPackages.otr</li>
    <li>ocamlPackages.prof_spacetime</li>
    <li>ocamlPackages.ptime</li>
    <li>ocamlPackages.tls</li>
    <li>ocamlPackages.vg</li>
    <li>ocamlPackages.virtual_dom</li>
    <li>ocamlPackages.webmachine</li>
    <li>ocamlPackages.x509</li>
    <li>ocamlPackages.xtmpl</li>
    <li>stog</li>
    <li>why3</li>
  </ul>
</details>

File: pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix [hunk 1]
```diff
@@ -1,28 +1,28 @@
...
-	src = fetchurl {
-		url = "https://github.com/ocsigen/js_of_ocaml/releases/download/${version}/js_of_ocaml-${version}.tbz";
-		sha256 = "0i0smhvsfx2ydmbyg5ai5cgqsfnng8rkcvys4i3fa55cw24aknrn";
-	};
+  src = fetchurl {
+    url = "https://github.com/ocsigen/js_of_ocaml/releases/download/${version}/js_of_ocaml-${version}.tbz";
+    sha256 = "069jyiayxcgwnips3adxb3d53mzd4rrq2783b9fgmsiyzm545lcy";
+  };
 
-	nativeBuildInputs = [ ocaml findlib dune_2 cppo menhir ];
```
Comment (vbgl) [inline pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix:16]: These three elements have been put in `nativeBuildInputs` in that PR: #73820. Probably for a reason. Isn’t this reason still valid?

ping @balsoft

Comment (balsoft) [reply pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix:16]: The reason is obvious -- all of those are `nativeBuildInputs` (they are build-time dependencies, so have to run on `build` platform) and not `buildInputs` (which would run on `host` platform).

File: pkgs/development/tools/ocaml/js_of_ocaml/camlp4.nix [hunk 2]
```diff
@@ -1,21 +1,19 @@
...
+    sha256 = "1v2hfq0ra9j07yz6pj6m03hrvgys4vmx0gclchv94yywpb2wc7ik";
+  };
 
-	inherit (js_of_ocaml-compiler) installPhase meta;
+  inherit (js_of_ocaml-compiler) meta useDune2;
 
-	buildInputs = [ ocaml findlib dune camlp4 ocsigen_deriving ];
-
-	buildPhase = "dune build -p js_of_ocaml-camlp4";
+  buildInputs = [ ocaml findlib dune camlp4 ocsigen_deriving ];
```
[the remark this answers is not in the document]
Comment (sternenseemann) [reply pkgs/development/tools/ocaml/js_of_ocaml/camlp4.nix:18]: No, I don't think so, missed them when cleaning up probably.

File: pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix (hunk 1 above)
Comment (sternenseemann) [reply pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix:16]: I was a bit too eager when deleting stuff here it seems. According to [js_of_ocaml-compiler's opam file](https://github.com/ocsigen/js_of_ocaml/blob/3.8.0/js_of_ocaml-compiler.opam) cppo is no longer required, but I'll re-add menhir.

Comment (sternenseemann) [reply pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix:16]: Nevermind what I wrote earlier, I wouldn't see why this made any difference or they would be required as `buildDunePackage` adds them to `buildInputs`. I think we don't support cross compiling for OCaml packages so this is no issue per se?!

Comment (vbgl) [reply pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix:16]: Thanks. But my question was more: is there a realistic scenario in which the host and build platforms are different?

Comment (sternenseemann) [reply pkgs/development/tools/ocaml/js_of_ocaml/compiler.nix:16]: I see. I'd say it'd be simpler to address that in buildDunePackage for the whole of `ocamlPackages`.

File: pkgs/development/tools/ocaml/js_of_ocaml/camlp4.nix [hunk 3]
```diff
@@ -1,21 +1,19 @@
...
-	};
+  src = fetchFromGitHub {
+    owner = "ocsigen";
+    repo = "js_of_ocaml";
+    rev = version;
+    sha256 = "1v2hfq0ra9j07yz6pj6m03hrvgys4vmx0gclchv94yywpb2wc7ik";
+  };
 
-	inherit (js_of_ocaml-compiler) installPhase meta;
+  inherit (js_of_ocaml-compiler) meta useDune2;
```
Comment (vbgl) [inline pkgs/development/tools/ocaml/js_of_ocaml/camlp4.nix:16]: Setting `useDune2` breaks this package.

Comment (sternenseemann) [reply pkgs/development/tools/ocaml/js_of_ocaml/camlp4.nix]: I've rebased to remove the inheritance of `useDune2` which seems to fix things.

**Pull request closed without merging**